sql/src/main/java/io/crate/operation/join/NestedLoopOperation.java

```diff
@@ -431,4 +376,351 @@ public void downstream(Projector downstream) {
+                            rowCombinator.combine(outerRow, innerRow)
+                    );
+                    rowsLeft--;
+                    wantMore = wantMore && rowsLeft > 0;
+                    if (!wantMore) {
+                        break Outer;
+                    }
+                }
+            }
+            return wantMore;
+        }
+
+        @Override
+        public TaskResult emptyResult() {
+            return PageableTaskResult.EMPTY_PAGEABLE_RESULT;
+        }
+
+        @Override
+        public int rowsToProduce(Optional<PageInfo> pageInfo) {
+            if (pageInfo.isPresent()) {
+                return pageInfo.get().position() + pageInfo.get().size();
+            } else {
+                return limit() + offset;
+            }
+        }
+
+        @Override
+        public void onFirstJoin(JoinContext joinContext) {
+            // do nothing
+        }
+
+        @Override
+        public TaskResult produceFirstResult(Object[][] rows, Optional<PageInfo> pageInfo, JoinContext joinContext) {
+            assert pageInfo.isPresent() : "pageInfo is not present for optimized paging";
+            IterableBigArray<Object[]> wrappedRows = new MultiNativeArrayBigArray<Object[]>(0, rows.length, rows);
+            return new NestedLoopPageableTaskResult(wrappedRows, 0L, pageInfo.get(), joinContext);
+        }
+
+        @Override
+        public String name() {
+            return "optimized paging";
+        }
+    }
+
+    /**
+     * fake paging that loads both relations to ram and pages through the
+     * projection results.
+     */
+    private class FakePagingNestedLoopStrategy extends DefaultNestedLoopStrategy {
+
+        @Override
+        public TaskResult emptyResult() {
+            return PageableTaskResult.EMPTY_PAGEABLE_RESULT;
+        }
+
+        @Override
+        public int rowsToProduce(Optional<PageInfo> pageInfo) {
+            return limit() + offset;
+        }
+
+        @Override
+        public void onFirstJoin(JoinContext joinContext) {
+            // we can close the context as we produced ALL results in one batch
+            try {
+                joinContext.close();
+            } catch (IOException e) {
+                logger.error("error closing joinContext after NestedLoop execution", e);
+            }
+        }
+
+        @Override
+        public TaskResult produceFirstResult(Object[][] rows, Optional<PageInfo> pageInfo, JoinContext joinContext) {
+            assert pageInfo.isPresent() : "pageInfo is not present for fake paging";
+            IterableBigArray<Object[]> wrappedRows = new MultiNativeArrayBigArray<Object[]>(0, rows.length, rows);
+            return new FetchedRowsPageableTaskResult(wrappedRows, 0L, pageInfo.get());
+        }
+
+        @Override
+        public String name() {
+            return "default paging";
+        }
+    }
+
+
+    private class DefaultNestedLoopStrategy implements NestedLoopStrategy {
+
```

Comment:
please name after what the strategy does here. that this is the default has nothing to do with what it does